[DEVEX-1339] Update EzLabel (#669)
ezcater/recipe: src/components/EzLabel/EzLabel.md

---
name: EzLabel

> title: Labelling

Author: `EzLabelledItem` title is also `Labelling` - do we want them to be the same?

Author: <img width="1182" alt="Screen Shot 2021-07-07 at 10 27 18 AM" src="https://user-images.githubusercontent.com/3790037/124812817-1a727680-df32-11eb-98ab-694d3b2f2e0c.png">

category: Inputs
path: '/components/ez-label'
tags: ['label']
---

Provides a label that can appear in one of several positions relative to its associated content. Typically used to label an interactive element or a short string of text.

---

## Best Practices

Labelled items should:

- Be used to label a static piece of text
- Be used with the `htmlFor` attribute to label form controls.

Labelled items should not:

> - Be used to label more than one child (unless being used with a single child that is a collection of multiple items, such as a Segmented Control)

Reviewer: Ah gotcha, thanks for the context.

I think `should only be associated with a single form input` is a good way to phrase this. We can definitely remove the Segmented Control text - it's more confusing than it is helpful.

> - Be used to label components that already contain labels, such as EzField. Use the label property of the component directly when possible e.g. `EzField[label={label}]`

Author: haha thanks for the credit - i didn't write that - it was from `EzLabelledItem`

---

## Content guidelines

Labelled items should

#### Do

- Use prop `use="primary"` to label important content, like form fields.
- Use prop `use="secondary"` to label supporting content, like a search field. Secondary labels are used to avoid drawing attention away from primary content.
- Use `EzLabel` without a `use` prop to label items without styling, in other words, with default body styling.

> - Use `htmlFor={controlId}` to associate the label with the form control. This allows the label to increase the hit-surface to focus/activate the input, as well as allowing assistive technology to identify the label when the input is focused.

Reviewer: ```suggestion
- Use `htmlFor={controlId}` to associate the label with the form control. This allows the label to increase the hit-surface to focus/activate the input, as well as allowing assistive technology to identify the label when the input is focused. Alternatively, an `<input>` element can be nested directly inside the label to automatically associate the label with the input element.
```

> 

Author: do we want to document the position options? do we want to document how to use `as='label'` and `as='div'`?

Reviewer: Yes, but probably not under the `do` section. The `do` section is intended as content guidelines - basically a means to communicate the _why_ behind the API. 

Perhaps if we change the `### Side label` example to `### Label positioning`, it'll give us a place we can talk about the other available options.

#### Don’t

- Don't use prop `use="secondary"` with `position="side"`

---

## Examples

### With input controls

Use `htmlFor={controlId}` to associate the label with the form control. This allows the label to increase the hit-surface to focus/activate the input, as well as allowing assistive technology to identify the label when the input is focused.

```jsx
<>
  <EzLabel htmlFor="search-input" use="secondary">
    Search
  </EzLabel>

>   <EzSearchInput id="search-input" placeholder="Search customers" />

Author: i don't love the extra margin above this search input, but it's not blocking

Author: <img width="1103" alt="Screen Shot 2021-07-07 at 10 57 49 AM" src="https://user-images.githubusercontent.com/3790037/124812868-29f1bf80-df32-11eb-9083-5fd15e2b070c.png">

Reviewer: I _think_ the extra margin here is coming from the card wrapping the example. Does the margin go away if you change the `<>` fragment wrapper for a `<div>`?

I would imagine the margin would also disappear if you did this::
```jsx
<EzLabel use="secondary">
  Search
  <EzSearchInput placeholder="Search customers" />
</EzLabel>
```

...but that's only if you want to demonstrate the other way you can associate a label with a form element (which we should probably also do).

</>
```

### Primary label

A label for important content.

```jsx
<>

>   <EzLabel use="primary">Primary Label</EzLabel>
>   Some text

Reviewer: I think we should be careful to demonstrate labels in ways we'd expect them to be used in downstream applications. A label without an associated input is problematic from an accessibility standpoint (given that it renders a `<label>` element).

Perhaps we could use an input here instead:
```suggestion
  <EzLabel use="primary" htmlFor="id="primary-input">Primary Label</EzLabel>
  <input id="primary-input" placeholder="Some text" />
```

or demonstrate the `as` prop - we'll have to be careful to specifically call out _why_ the `as` prop is needed:
```suggestion
  <EzLabel use="primary" as="span">Primary Label</EzLabel>
  Some text
```

</>
```

### Secondary label

This is alternative style of top label. Use this label style when labelling controls inside the page header or subheader. This label style should also be used to provide additional context in an understated way, as not to draw the focus away from important data.

```jsx
<>
  <EzLabel use="secondary">Secondary Label</EzLabel>
  Some text
</>
```

### Side label

A label that appears next to its content.

```jsx
<>
  <EzLabel position="side" use="primary">
    Side Primary Label
  </EzLabel>
  Some text
</>
```

> ### Bottom label

Reviewer: Thought's on killing the `### Bottom label` example and combining it with `### Side label` as `### Label positioning`?

A label that appears below its content.

```jsx
<>
  Some text

>   <EzLabel use="primary">Bottom Primary Label</EzLabel>

Reviewer: Is this missing `position="bottom"`?

Author: does it need it? it's already below it?

```jsx
<EzLabel htmlFor="search-input" use="secondary">
  Search
  <EzSearchInput id="search-input" placeholder="Search customers" />
</EzLabel>
```

### Primary label

A label for important content.

```jsx
<div>
  <EzLabel use="primary" htmlFor="primary-input">
    Primary Label
  </EzLabel>
  <input id="primary-input" placeholder="Some text" />
</div>
```

### Secondary label

This is alternative style of top label. Use this label style when labelling controls inside the page header or subheader. This label style should also be used to provide additional context in an understated way, as not to draw the focus away from important data.

```jsx
<div>
  <EzLabel use="secondary">Secondary Label</EzLabel>
  Some text
</div>
```

### Labels with default body styling

A `EzLabel` without the `use` prop label with default body styling.

```jsx
<div>
  <EzLabel htmlFor="text-container">Secondary Label</EzLabel>
  <div id="text-container">Some text</div>
</div>
```

### Label positioning

> A label that appears next to its content.

Reviewer: ```suggestion
Top labels are the default position for `EzLabel` and are recommended because they work better with long copy, localization, and responsive layouts. Side labels can be useful when vertical space is limited.

The label's overall position relative to the element it is labeling may be changed using the `position` prop.
```

```jsx
<div>
  <EzLabel position="side" use="primary">
    Side Primary Label
  </EzLabel>

>   Some text

Reviewer: Let's use an input here.

</div>
```

> A label that appears below its content.

Reviewer: Let's skip the `position=bottom` example.

```jsx
<div>
  Some text
  <EzLabel use="primary" position="bottom">
    Bottom Primary Label
  </EzLabel>
</div>
```

Hidden labels, with `position='hidden'`, are used to programmatically associate an input element with a label for accessibility purposes.

```jsx
<div>

>   <EzLabel position="hidden" htmlFor="hidden-content">

Reviewer: Nitpick: it's not the content that is hidden, it's the label, so this `Id` value is misleading.

Specifically we're trying to show how folks can use `EzLabel` to give assistive technology extra cues about the purpose of a given input element when the label would otherwise be visually hidden. Typically you'll find labels are omitted from inputs when the context is otherwise clear from the page structure/layout - something that may not be so clear for visually impaired users.

    Label
  </EzLabel>
  <div id="hidden-content">Some text</div>
</div>
```

### Labels with error state

When provided the `error` prop, the label will receive error styling.

```jsx
<div>
  <EzLabel use="primary" error={true}>
    Label
  </EzLabel>

>   Some text

Reviewer: Please use an input here too.

</div>
```

### Label accessibility

> Labels render `<label>` elements by default, and with the `as='label'` prop, which provide accessibility behaviors. To render an element that looks like a label without accessibility behaviors, provide another element to the `EzLabel` prop `as`, examples: `as='div'` or `as='span'`.

Reviewer: ```suggestion
Labels render `<label>` elements by default, allowing form elements to be programmatically associated with an identifying text caption and allows browsers to provide additional user experience benefits such as using the label to provide a larger hit-target for inputs.

To render an element that looks like a label without rendering a html `<label>` element, the `as` prop can be provided with another element type, such as: `as='div'` or `as='span'`.

This may be useful if providing a top-level label for a group of inputs where each individual input may already have a label, such as a radio button group.

NOTE: the `htmlFor` prop should be omitted when using the `as` prop. In order to retain programmatic association of the label with the content, additional `aria-` attributes such as `aria-labelledby` may be necessary.
```
